feat: Add DPoP (Demonstrating Proof of Possession) authentication support (#183)
auth0/node-oauth2-jwt-bearer: packages/express-oauth2-jwt-bearer/EXAMPLES.md

# Examples

- [DPoP Authentication (Early Access)](#dpop-authentication-early-access)
  - [Accept both Bearer and DPoP tokens (default)](#accept-both-bearer-and-dpop-tokens-default)
  - [Require only DPoP tokens](#require-only-dpop-tokens)
  - [Require only Bearer tokens](#require-only-bearer-tokens)

>   - [Customize DPoP validation behavior](#customize-dpop-validation-behavior)

Reviewer: This is missing the `Hostname Resolution (req.host and req.protocol)` heading.

  
- [Restrict access with scopes](#restrict-access-with-scopes)
- [Restrict access with claims](#restrict-access-with-claims)
  - [Matching a specific value](#matching-a-specific-value)
  - [Matching multiple values](#matching-multiple-values)
  - [Matching custom logic](#matching-custom-logic)

## DPoP Authentication (Early Access) 

> >**Note:** DPoP (Demonstration of Proof-of-Possession) support is currently in [**Early Access**](https://auth0.com/docs/troubleshoot/product-lifecycle/product-release-stages#early-access). Contact Auth0 support to enable it for your tenant.

Reviewer: ```suggestion
> [!NOTE]  
> DPoP (Demonstration of Proof-of-Possession) support is currently in [**Early Access**](https://auth0.com/docs/troubleshoot/product-lifecycle/product-release-stages#early-access). Contact Auth0 support to enable it for your tenant.
>
> If DPoP is disabled (`dpop: { enabled: false }`), only standard Bearer tokens will be accepted.
```

> If DPoP is disabled (`dpop: { enabled: false }`), only standard Bearer tokens will be accepted.

> By default, DPoP is supported but optional, meaning the middleware can accept both Bearer and DPoP tokens. See the examples below to configure strict or relaxed behavior.

Author: Updated here: https://github.com/auth0/node-oauth2-jwt-bearer/pull/183/commits/2f6096f0e67f7a7a7c04603153160a3b61655378#diff-5d20390a3127d7a69396204dc73d52cf084f9e5c1c1ae50f1f45bcb8fc25a523L18

### Accept both Bearer and DPoP tokens (default)
```js
const { auth } = require('express-oauth2-jwt-bearer');

app.use(
  auth({
    issuerBaseURL: 'https://YOUR_ISSUER_DOMAIN',
    audience: 'https://my-api.com',
    dpop: {
      enabled: true,   // Enables DPoP support
      required: false  // Accepts both Bearer and DPoP tokens (default)
    }
  })
);

app.get('/api/resource', (req, res) => {
  res.send('Access granted');
});
```
Requests using DPoP must include both headers:
```http
Authorization: DPoP eyJhbGciOiJFUzI1NiIsInR5cCI6...
DPoP: eyJhbGciOiJkcG9wIiwidHlwIjoi...
```

### Require only DPoP tokens
To enforce stronger protection and reject non-DPoP tokens:
```js
const { auth } = require('express-oauth2-jwt-bearer');

app.use(
  auth({
    issuerBaseURL: 'https://YOUR_ISSUER_DOMAIN',
    audience: 'https://my-api.com',
    dpop: {
      enabled: true,
      required: true  // Rejects Bearer tokens
    }
  })
);

app.get('/api/secure-resource', (req, res) => {
  res.send('DPoP token validated');
});
```

### Require only Bearer tokens

> If you want to reject all DPoP-bound tokens and only accept standard Bearer access tokens, you can disable DPoP support explicitly:

Reviewer: ```suggestion
If you want to reject all DPoP tokens and only accept standard Bearer access tokens, you can disable DPoP support explicitly:
```

Author: Updated here: https://github.com/auth0/node-oauth2-jwt-bearer/pull/183/commits/2f6096f0e67f7a7a7c04603153160a3b61655378#diff-5d20390a3127d7a69396204dc73d52cf084f9e5c1c1ae50f1f45bcb8fc25a523R72

```js
const { auth } = require('express-oauth2-jwt-bearer');

app.use(
  auth({
    issuerBaseURL: 'https://YOUR_ISSUER_DOMAIN',
    audience: 'https://my-api.com',
    dpop: {
      enabled: false  // DPoP proofs will be ignored
    }
  })
);

app.get('/api/bearer-only', (req, res) => {
  res.send('This route accepts only Bearer tokens');
});
```

### Customize DPoP validation behavior
```js
const { auth } = require('express-oauth2-jwt-bearer');

app.use(
  auth({
    issuerBaseURL: 'https://YOUR_ISSUER_DOMAIN',
    audience: 'https://my-api.com',
    dpop: {
      enabled: true,
      required: true,
      iatOffset: 300,   // Reject proofs older than 5 minutes
      iatLeeway: 30,    // Allow 30s clock skew
    }
  })
);
```
#### DPoP Behavior Matrix

| `enabled` | `required` | Behavior                                                                                             |
| --------- | ---------- | -----------------------------------------------------------------------------------------------------|

> | `true`    | `false`    | **Default behavior**.Both Bearer and DPoP tokens are accepted. Proofs are validated if present.      |

Reviewer: ```suggestion
| `true`    | `false`    | **Default behavior**. Both Bearer and DPoP tokens are accepted. Proofs are validated if present.      |
```

Author: Updated here: https://github.com/auth0/node-oauth2-jwt-bearer/pull/183/commits/2f6096f0e67f7a7a7c04603153160a3b61655378#diff-5d20390a3127d7a69396204dc73d52cf084f9e5c1c1ae50f1f45bcb8fc25a523L110

| `false`   | `false`    | Only Bearer tokens are accepted. DPoP headers are ignored.                                           |
| `false`   | `true`     | Invalid configuration. DPoP is ignored, so `required: true` has no effect. DPoP is ignored entirely. |

> | `true`    | `true`     | Only DPoP-bound tokens are accepted. Bearer tokens are rejected.                                     |

Reviewer: ```suggestion
| `true`    | `true`     | Only DPoP tokens are accepted. Bearer tokens are rejected.                                     |
```

Author: Updated here: https://github.com/auth0/node-oauth2-jwt-bearer/pull/183/commits/2f6096f0e67f7a7a7c04603153160a3b61655378#diff-5d20390a3127d7a69396204dc73d52cf084f9e5c1c1ae50f1f45bcb8fc25a523L112

#### Proof Timing Options

When DPoP is enabled, you can control the accepted timing of DPoP proofs using the following options:

  - `iatOffset`: The maximum age (in seconds) of a DPoP proof. Proofs with `iat` older than this offset (relative to now) will be rejected.
    Default: `300 seconds`(5 minutes)

>   - `iatLeeway`: Clock skew tolerance (in seconds) when comparing proof `iat` with the current server time.

Reviewer: ```suggestion
  - `iatLeeway`: Clock skew tolerance (in seconds) when comparing a proof's `iat` with the current server time.
```

    Default: `30 seconds`

### Hostname Resolution (`req.host` and `req.protocol`)

> This SDK uses `req.protocol` and `req.host` to construct the htu (HTTP URI) claim for validating DPoP proofs.

Author: Updated here: https://github.com/auth0/node-oauth2-jwt-bearer/pull/183/commits/2f6096f0e67f7a7a7c04603153160a3b61655378#diff-5d20390a3127d7a69396204dc73d52cf084f9e5c1c1ae50f1f45bcb8fc25a523R130
